crawl-ref/source/util/tag-35-upgrade.py

```diff
@@ -0,0 +1,387 @@
+import os
+from typing import Sequence, Dict
+
+logging.basicConfig(level=logging.DEBUG)
+
+NEW_TAG = 35
+
+
+def _edit_file(
+    file_name: str, delete: Sequence[str] = [], replace: Dict[str, str] = {}
+):
+    with open(file_name) as f:
+        lines = f.readlines()
+
+    # Sanity checks for unhandled modifications of source file
+    for line in delete:
+        if line not in lines:
+            raise ValueError(f"Asked to delete this line but couldn't find it: {line}")
+    for line in replace.keys():
+        if line not in lines:
+            raise ValueError(f"Asked to replace this line but couldn't find it: {line}")
+
+    logging.info(
+        "Updating %s (%s deletions, %s modifications)",
+        file_name,
+        len(delete),
+        len(replace),
+    )
+    with open(file_name, "w") as f:
+        f.writelines(replace.get(l, l) for l in lines if l not in delete)
+
+
+def _delete_file(file_name):
+    if not os.path.isfile(file_name):
+        raise ValueError(f"Asked to delete non-existent file {file_name}")
+    logging.info("Deleting %s", file_name)
+    os.unlink(file_name)
+
+
+def _trim_file(file_name, delete_marker):
+    with open(file_name) as f:
+        lines = f.readlines()
+    last_line = lines.index(delete_marker)
+    logging.info(
+        "Trimming the last %s lines from %s", len(lines) - last_line - 1, file_name
+    )
+    with open(file_name, "w") as f:
+        f.writelines(lines[: last_line - 1])
+
+
+def _snip_file(file_name, start_snip, end_snip):
+    with open(file_name) as f:
+        lines = f.readlines()
+    first_line = lines.index(start_snip)
+    last_line = first_line + lines[first_line:].index(end_snip) + 1
+    num_lines = last_line - first_line
+    logging.info(
+        "Snipping %s lines (%s to %s) from %s",
+        num_lines,
+        first_line,
+        last_line,
+        file_name,
+    )
+    for _ in range(num_lines):
+        del lines[first_line]
+    with open(file_name, "w") as f:
+        f.writelines(lines)
+
+
+def main():
+    species()
+    art_data_txt()
+    art_data_pl()
+    rltiles()
+    des()
+
+
+def species():
+    SPECIES_BASE = "dat/species"
+    for name in (
+        "barachi",
+        "centaur",
+        "deep-dwarf",
+        "deep-elf",
+        "demigod",
+        "demonspawn",
+        "deprecated-draconian-mottled",
+        "draconian-base",
+        "draconian-black",
+        "draconian-green",
+        "draconian-grey",
+        "draconian-pale",
+        "draconian-purple",
+        "draconian-red",
+        "draconian-white",
+        "draconian-yellow",
+        "felid",
+        "formicid",
+        "gargoyle",
+        "ghoul",
+        "halfling",
+        "hill-orc",
+        "human",
+        "kobold",
+        "merfolk",
+        "minotaur",
+        "mummy",
+        "naga",
+        "octopode",
+        "ogre",
+        "spriggan",
+        "tengu",
+        "troll",
+        "vampire",
+        "vine-stalker",
+    ):
```

Comment:
maybe use `os.walk`?

Reply:
It lacks gnoll or any new species that will get added in future.

Reply:
I did this 😊